### PR TITLE
Widgets for saving and loading distances

### PR DESCRIPTION
**Pull request by janezd**
The PR contains

- methods for loading and saving to `DistMatrix` from/to file
- the corresponding widgets
- some additional tests
- removal of `DistMatrix.get_KNN`

It also pulls the code for recent paths from the file widget into a separate module, so it can be reused in multiple widgets.

Widgets for file distances intentionally support just a single data format. Let's try to keep it simple this time.

### CODE REVIEW

File: Orange/tests/test_distances.py [hunk 1]
```diff
@@ -34,6 +37,157 @@ def test_pickling(self):
...
+    def named_file(self, content, encoding=None):
+        import tempfile
+        import os
+        file = tempfile.NamedTemporaryFile(
+                "wt", delete=False, encoding=encoding)
+        file.write(content)
+        name = file.name
+        file.close()
+        yield name
+        os.remove(name)
```
Comment (ales-erjavec) [inline Orange/tests/test_distances.py]: this should be
```python
try:
    yield name
finally:
    os.remove(name)
```

File: Orange/widgets/unsupervised/owdistancefile.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                start_file = ""
+            if not start_file or not os.path.exists(start_file):
+                widgets_dir = os.path.dirname(gui.__file__)
+                orange_dir = os.path.dirname(widgets_dir)
+                start_file = os.path.join(orange_dir, "doc", "datasets")
+            if not start_file or not os.path.exists(start_file):
+                d = os.getcwd()
+                if os.path.basename(d) == "canvas":
+                    d = os.path.dirname(d)
+                start_file = os.path.join(os.path.dirname(d), "doc", "datasets")
```
Comment (ales-erjavec) [inline Orange/widgets/unsupervised/owdistancefile.py]: All of this (after `get_sample_datasets_dir()`) is unnecessary (the same applies to the OWFile widget also).
From memory this is how Orange was organized in v2 (possibly even v1).

File: Orange/tests/test_distances.py (hunk 1 above)
Comment (lanzagar) [inline Orange/tests/test_distances.py]: Should we have a contextmanager like this implemented properly and available in some utils module?
We appear to be using something similar in multiple places in Orange (search "NamedTemporaryFile")

File: Orange/widgets/utils/filedialogs.py [hunk 3]
```diff
@@ -85,3 +87,269 @@ def get_file_name(start_dir, start_filter, file_formats):
...
+        """Move the n-th file to the top of the list"""
+        recent = self.recent_paths[n]
+        del self.recent_paths[n]
+        self.recent_paths.insert(0, recent)
+
+    def last_path(self):
+        """Return the most recent absolute path or `None` if there is none"""
+        abspath = self.recent_paths and self.recent_paths[0].abspath
+        if abspath != "(none)":
+            return abspath
```
Comment (ales-erjavec) [inline Orange/widgets/utils/filedialogs.py]: if `self.recent_paths` is `[]` (the default value) this will return an `[]` not a `None`. Also the abspath of the recent item should never be a string "(none)".

File: Orange/widgets/data/owfile.py [hunk 4]
```diff
@@ -21,128 +22,7 @@ def add_origin(examples, filename):
             var.attributes["origin"] = dir_name
 
 
-class RecentPath:
```
Comment (ales-erjavec) [inline Orange/widgets/data/owfile.py]: Please import RecentPath item from the new location into this module's namespace so that settings saved before the change can be loaded.

File: Orange/widgets/utils/filedialogs.py [hunk 5]
```diff
@@ -85,3 +87,269 @@ def get_file_name(start_dir, start_filter, file_formats):
...
+        return None
+
+    @property
+    def value(self):
+        return os.path.basename(self.abspath)
+
+    @property
+    def icon(self):
+        provider = QFileIconProvider()
+        return provider.icon(QFileIconProvider.Drive)
```
Comment (ales-erjavec) [inline Orange/widgets/utils/filedialogs.py]: `return provider.icon(QFileInfo(self.abspath))`?

File: Orange/widgets/utils/filedialogs.py [hunk 6]
```diff
@@ -85,3 +87,269 @@ def get_file_name(start_dir, start_filter, file_formats):
...
+                if self.prefix == prefix:
+                    path = os.path.join(base, self.relpath)
+                    if os.path.exists(path):
+                        return RecentPath(
+                            os.path.normpath(path), self.prefix, self.relpath)
+        return None
+
+    @property
+    def value(self):
+        return os.path.basename(self.abspath)
```
Comment (ales-erjavec) [inline Orange/widgets/utils/filedialogs.py]: Is `value` being used anywhere? I would prefer it it was renamed to basename.

File: Orange/tests/test_distances.py (hunk 1 above)
Comment (janezd) [inline Orange/tests/test_distances.py]: I moved this context manager to `Orange.tests.named_file`. I haven't looked for all other uses; this belongs to another PR.

File: Orange/widgets/unsupervised/owdistancefile.py (hunk 2 above)
Comment (janezd) [inline Orange/widgets/unsupervised/owdistancefile.py]: OK, I also removed it from the File Widget. Please check whether I understood it correctly.

File: Orange/widgets/utils/filedialogs.py (hunk 6 above)
Comment (janezd) [inline Orange/widgets/utils/filedialogs.py]: Renamed to `basename`. It was used in two places.

**Comment by janezd** [comment]
@ales-erjavec Is this OK now?
